cloud/smartdatacenter/sdc.py

```diff
@@ -0,0 +1,318 @@
+    location: 'us-east-1.api.joyentcloud.com'
+    count_tag:
+      role: webserver
+    exact_count: 5
+
+# Delete machines by their tags
+- local_action:
+    module: sdc
+    account: myaccount
+    key_id: /myaccount/keys/mykey
+    location: 'us-east-1.api.joyentcloud.com'
+    tags:
+      role: webserver
+    state: absent
+
+# Delete machine by its id
+- local_action:
+    module: sdc
+    account: myaccount
+    key_id: /myaccount/keys/mykey
+    machine_id: 51307f2f-02a3-4bdc-9b38-2f263eb43e2d
+    location: 'us-east-1.api.joyentcloud.com'
+    state: absent
+'''
+
+import os
+import sys
+import time
+try:
+    import smartdc
+    from smartdc import DataCenter
+    HAS_SMARTDC = True
+except ImportError:
+    HAS_SMARTDC = False
+
+def _wait_for_status(module, machines, status):
+    wait_timeout = int(module.params.get('wait_timeout'))
+    wait_timeout_time = time.time() + wait_timeout
+    while not all( sm.status() == status for sm in machines ) and wait_timeout_time > time.time():
+        time.sleep(2)
+    # Check again that all are in expected state. If not, we timed out
+    if not all( sm.status() == status for sm in machines ):
+        module.fail_json(msg="Timed out creating machines")
+
+def list_existing_machines(module, sdc, tags):
+    if not tags:
+        module.fail_json(msg="Must have tags to list existing machines for deletion/counting")
+    machines = sdc.machines(tags=tags, state='running')
+    return machines
+
+
+def create_virtual_machine(module, sdc):
+    """
+    Create new virtual machine
+
+    module : AnsibleModule object
+    sdc: authenticated sdc Datacenter object
+
+    Returns:
+        True if a new machine was created, false otherwise
+    """
+    name = module.params.get('name')
+    package = module.params.get('package')
+    image = module.params.get('image')
+    networks = module.params.get('networks')
+    wait = module.params.get('wait')
+    wait_timeout = int(module.params.get('wait_timeout'))
+    tags = module.params.get('tags')
+    exact_count = int(module.params.get('exact_count'))
+    count_tag = module.params.get('count_tag')
+    count = int(module.params.get('count'))
+
+    changed = False
+    created_machines = []
+    existing_machines = []
+    full_tags = dict(tags.items() + count_tag.items()) #Force count_tag(s) to be a part of the instance tags to prevent user error
+
+    # Get existing machines if we're using exact_count
+    if count_tag and exact_count:
+        existing_machines = list_existing_machines(module, sdc, count_tag)
+        count_needed = exact_count - len(existing_machines)
+    else:
+        count_needed = count
+
+    #Make sure we actually need to create machines, otherwise we can just give info about the existing ones
+    if count_needed > 0:
+        changed = True
+        # Batch create, we'll wait later
+        for x in range(count_needed):
+            sm = sdc.create_machine(name=name, package=package, image=image,
```

Comment:
no exception handling? I expect sdc to throw exceptions for several reasons, they should be captured and returned through i.e. fail_json(msg=nice message, trackeback=exception)

the same for the other calls to sdc